exec/java-exec/src/main/java/org/apache/drill/exec/store/avro/AvroBatchReader.java

```diff
@@ -0,0 +1,368 @@
+      String fieldName = field.name();
+      Object value = record.get(fieldName);
+      ObjectWriter writer = rowWriter.column(fieldName);
+      processRecord(writer, value, field.schema());
+    }
+    rowWriter.save();
+    return true;
+  }
+
+  private void processRecord(ObjectWriter writer, Object value, Schema schema) {
+    // skip processing record if it is null or is not projected
+    if (value == null || !writer.isProjected()) {
+      return;
+    }
+
+    switch (schema.getType()) {
+      case UNION:
+        processRecord(writer, value, AvroSchemaUtil.extractSchemaFromNullable(schema, writer.schema().name()));
+        break;
+      case RECORD:
+        TupleWriter tupleWriter = writer.tuple();
+
+        if (tupleWriter.tupleSchema().isEmpty()) {
+          // fill in tuple schema for cases when there recursive named record types are present
+          TupleMetadata recordSchema = AvroSchemaUtil.convert(schema);
+          recordSchema.toMetadataList().forEach(tupleWriter::addColumn);
+        }
+
+        GenericRecord genericRecord = (GenericRecord) value;
+        schema.getFields().forEach(
+          field -> processRecord(tupleWriter.column(field.name()), genericRecord.get(field.name()), field.schema())
+        );
+        break;
+      case ARRAY:
+        ArrayWriter arrayWriter = writer.array();
+        GenericArray<?> array = (GenericArray<?>) value;
+        ObjectWriter entryWriter = arrayWriter.entry();
+        for (Object arrayValue : array) {
+          processRecord(entryWriter, arrayValue, array.getSchema().getElementType());
+          arrayWriter.save();
+        }
+        break;
+      case MAP:
+        @SuppressWarnings("unchecked")
+        Map<Object, Object> map = (Map<Object, Object>) value;
+        Schema valueSchema = schema.getValueType();
+
+        DictWriter dictWriter = writer.dict();
+        ScalarWriter keyWriter = dictWriter.keyWriter();
+        ObjectWriter valueWriter = dictWriter.valueWriter();
+
+        for (Map.Entry<Object, Object> mapEntry : map.entrySet()) {
+          processScalar(keyWriter, mapEntry.getKey());
+          processRecord(valueWriter, mapEntry.getValue(), valueSchema);
+          dictWriter.save();
+        }
+        break;
+      default:
+        try {
+          ScalarWriter scalarWriter = writer.scalar();
+          processScalar(scalarWriter, value);
+        } catch (UnsupportedOperationException e) {
+          throw UserException.dataReadError(e)
+            .message("Unexpected writer type '%s', expected scalar", writer.type())
+            .addContext("Reader", this)
+            .build(logger);
+        }
+    }
+  }
+
+  private void processScalar(ScalarWriter scalarWriter, Object value) {
+    ColumnMetadata columnMetadata = scalarWriter.schema();
+    switch (columnMetadata.type()) {
+      case INT:
+        scalarWriter.setInt((int) value);
+        break;
+      case BIGINT:
+        scalarWriter.setLong((long) value);
```

Comment:
Where the Avro-provided Java object is of the type that the scalar writer wants, we can just call `setValue(Object value)` which will do the required casting for you.

Reply:
Well, most of the scalar type still have their own logic so left those that can be replaces with setObject with casts for visibility. Having some casts implicitly, others not might be a little confusing.